File: Mandelbrot-Set/Javascript/2.js
const { createCanvas } = require('canvas');

function mandelbrot(c_real, c_imag, max_iterations) {
  let z_real = 0;
  let z_imag = 0;
  let n = 0;

  while (z_real * z_real + z_imag * z_imag <= 4 && n < max_iterations) {
    const newReal = z_real * z_real - z_imag * z_imag + c_real;
    const newImag = 2 * z_real * z_imag + c_imag;
    z_real = newReal;
    z_imag = newImag;
    n++;
  }

  return n;
}

function create_mandelbrot_set(width, height, x_min, x_max, y_min, y_max, max_iterations) {
  const canvas = createCanvas(width, height);
  const ctx = canvas.getContext('2d');

  const imageData = ctx.createImageData(width, height);
  const data = imageData.data;

  let dataIndex = 0;

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const real = x_min + (x / (width - 1)) * (x_max - x_min);
      const imag = y_min + (y / (height - 1)) * (y_max - y_min);

      const iteration = mandelbrot(real, imag, max_iterations);

      // Map the number of iterations to a color
      const r = iteration % 256;
      const g = (iteration * 2) % 256;
      const b = (iteration * 5) % 256;

      data[dataIndex] = r;
      data[dataIndex + 1] = g;
      data[dataIndex + 2] = b;
      data[dataIndex + 3] = 255; // Set alpha to 255 (fully opaque)

      dataIndex += 4;
    }
  }

  ctx.putImageData(imageData, 0, 0);

  return canvas;
}

function main() {
  const WIDTH = 3200;
  const HEIGHT = 3200;
  const X_MIN = -2.5;
  const X_MAX = 1.5;
  const Y_MIN = -2.0;
  const Y_MAX = 2.0;
  const MAX_ITERATIONS = 5000;
  
  const mandelbrot_canvas = create_mandelbrot_set(WIDTH, HEIGHT, X_MIN, X_MAX, Y_MIN, Y_MAX, MAX_ITERATIONS);
  const fs = require('fs');
  const out = fs.createWriteStream(__dirname + '/mandelbrot.png');
  const stream = mandelbrot_canvas.createPNGStream();
  stream.pipe(out);
  out.on('finish', () => console.log('The PNG file was created.'));
}

const start = process.hrtime.bigint();
main()
const end = process.hrtime.bigint();
const duration = end - start;
console.log(`Execution time: ${duration} ns`);

File: Mandelbrot-Set/Python/2.py
# Uses numpy so arrays are actually C/C++
import numpy as np
import time
from PIL import Image

def mandelbrot_numpy(c, max_iterations):
    z = np.zeros_like(c, dtype=complex)
    n = np.zeros(c.shape, dtype=int)

    for i in range(max_iterations):
        # Bool array representing if the point is part or isnt part of the mandelbrot set (bonded or goes to infinity)
        mask = np.less(np.abs(z), 2)
        z[mask] = z[mask] * z[mask] + c[mask]
        n[mask] = i
    return n

def create_mandelbrot_set(width, height, x_min, x_max, y_min, y_max, max_iterations):
    # Create a grid of complex numbers covering a rectangle
    x, y = np.ogrid[y_min:y_max:height*1j, x_min:x_max:width*1j]
    c = x + y * 1j
    mandelbrot = mandelbrot_numpy(c, max_iterations)

    # Assign colors based on iteration number
    colors = (mandelbrot % 256, (mandelbrot * 2) % 256, (mandelbrot * 5) % 256)
    # Transpose is needed because np is row-major and img libraries have colors on the last dimension (RGB)
    colors = np.transpose(colors, axes=(1, 2, 0))

    return Image.fromarray(colors.astype('uint8'), 'RGB')

def main():
    WIDTH, HEIGHT = 800, 800
    X_MIN, X_MAX = -2.5, 1.5
    Y_MIN, Y_MAX = -2.0, 2.0
    MAX_ITERATIONS = 1000

    mandelbrot_image = create_mandelbrot_set(WIDTH, HEIGHT, X_MIN, X_MAX, Y_MIN, Y_MAX, MAX_ITERATIONS)
    mandelbrot_image.save("mandelbrot.png")

if __name__ == "__main__":
    start = time.time_ns()
    main()
    duration = time.time_ns() - start
    print(f"Execution time: {duration} ns")
